security/l1-contracts-dev-flow-vision.md

```diff
@@ -0,0 +1,159 @@
+# Purpose
+
+The goal of this design doc is to outline the vision we are working towards for the L1 contracts
+development flow, so we can collect feedback from stakeholders on what we are working towards to
+ensure we are all aligned on the vision.
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+The EVM Safety team is working on two projects—OPCM (OP Contracts Managr) Upgrades and superchain-ops
+improvements—that aim to improve the tools and standardize the processes around the full L1 smart
+contract development cycle. This includes development, testing, releasing, deployment, and upgrading.
+
+# Problem Statement + Context
+
+<!-- Describe the specific problem that the document is seeking to address as well
+as information needed to understand the problem and design space.
+If more information is needed on the costs of the problem,
+this is a good place to that information. -->
+
+As the number of chains we need to manage increases, our existing L1 contract development processes
+have proven to be insufficient, and will not scale well. We need to tighten up this end to end flow
+with robust, easy to use tools and well-defined processes. Some existing issues we've had are:
+
+- Deployment and upgrade processes are not standardized.
+- Confusion around smart contract releases.
+- Calldata and state changes are not part of governance proposals and not verifiable by governance.
+- Task development and validation is a manual process that is time-consuming and error-prone.
+- We do not have a good way to track the full changelog for a given contracts release.
+
+# Proposed Solution
+
+<!-- A high level overview of the proposed solution.
+When there are multiple alternatives there should be an explanation
+of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API)
+is likely too low level. -->
+
+The proposed UX upon for L1 smart contract development is described below.
+This UX will be achieved once the ongoing OPCM Upgrades and superchain-ops projects are completed.
+We start from assuming that `op-contracts/v2.0.0` with OPCM Upgrades is shipped as part of Isthmus
+in Cycle 34. We intentionally exclude the upgrade to `op-contracts/v2.0.0` in this section, as it
+may have some unique aspects because it puts this system in place, although we expect it to be
+largely similar to the flow described here.
+
+## Development
+
+A team starts working on a feature. If, and only if, they are 100% sure this feature is going
+into the next release, feature work can happen on contracts directly. Otherwise the feature must be
+developed using the inheritance pattern described in [`smart-contract-feature-development.md`](../smart-contract-feature-development.md).
```

Comment:
I would express this as requiring the tip of `develop` to always be releasable. You can never be 100% sure that a change will go into the next release unless it is merged as a single PR because you never know when an emergency release may be required or something else happens that changes the plan. So basically stuff that's small enough to ship to production with a single PR can just be done, everything else needs some form of feature toggle (the inheritance pattern).